### PR TITLE
[Resolve #1291] Fix attribute error caused by a relative project path

### LINKED ISSUE #1291: Intermittent AttributeError when resolving dependencies using the sceptre.plan.plan module.
### Subject of the issue

Hello,

I've got an intermittent AttributeError when resolving dependencies using the sceptre.plan.plan module. It only occurs when using a particular hierarchy of dependencies. The error does not occur when using equivalent Sceptre CLI commands. I'm not sure if there's something I'm overlooking in my usage of the API. My config appears to be consistent with the CLI. I didn't spot anything similar when looking through existing issues regarding dependencies.

Cheers,

Pat

The dependency hierarchy:
- Stack 1 depends on Stack 2 and Stack 3.
- Stack 2 depends on Stack 3.

Workarounds:
- Removing either Stack 1s dependency on Stack 2, or Stack 2s dependency on Stack 3 stops the issue occurring.
- Adding `ignore_dependencies` to the SceptreContext will prevent the error. This seems to be practical for what I need to do.

The trace:
```
Traceback (most recent call last):
  File "/home/pjls/repos/sceptre-bug/sceptre-describe.py", line 10, in <module>
    change_sets = plan.describe()
  File "/home/pjls/.local/lib/python3.9/site-packages/sceptre/plan/plan.py", line 208, in describe
    self.resolve(command=self.describe.__name__)
  File "/home/pjls/.local/lib/python3.9/site-packages/sceptre/plan/plan.py", line 114, in resolve
    self.launch_order = self._generate_launch_order(reverse)
  File "/home/pjls/.local/lib/python3.9/site-packages/sceptre/plan/plan.py", line 62, in _generate_launch_order
    graph = self.graph.filtered(self.command_stacks, reverse)
  File "/home/pjls/.local/lib/python3.9/site-packages/sceptre/config/graph.py", line 47, in filtered
    relevant |= nx.algorithms.dag.ancestors(graph, stack)
  File "/home/pjls/.local/lib/python3.9/site-packages/sceptre/stack.py", line 251, in __eq__
    self.dependencies == stack.dependencies and
  File "/home/pjls/.local/lib/python3.9/site-packages/sceptre/stack.py", line 242, in __eq__
    self.name == stack.name and
AttributeError: 'str' object has no attribute 'name'
```

### Your environment
* version of sceptre (sceptre --version): 3.2.0
* version of python (python --version): 3.9.2
* which OS/distro: Debian Bullseye

### Steps to reproduce
Using the sample repository here: https://github.com/pjsmith404/sceptre-bug

1. Create the stacks using `sceptre launch .`
2. Create a change set using `sceptre create experimental test-change-set`
3. Describe the change set using `sceptre describe change-set experimental test-change-set
4. Describe the change set using `python sceptre-describe.py`. Sometimes it will work, most times it will throw the AttributeError

### Expected behaviour
I would expect dependency resolution via the CLI or the API to return the same result.

### Actual behaviour
Dependency resolution via the API causes intermittent AttributeErrors

### PR DESCRIPTION
**Pull request by pjsmith404**
Resolve [#1291](https://github.com/Sceptre/sceptre/issues/1291)

Given a relative project path with a leading `.` and duplicate dependencies in multiple stack configurations, there is a chance of an AttributeError being raised. This is caused by dependencies failing to resolve into a valid Stack object due to an invalid path.

This PR fixes the issue by ensuring a given project path is resolved to the absolute path when initialising the Context.

An additional test of context with a relative path is included to ensure correct resolution.

## PR Checklist

- [x] Wrote a good commit message & description [see guide below].
- [x] Commit message starts with `[Resolve #issue-number]`.
- [x] Added/Updated unit tests.
- [x] Added/Updated integration tests (if applicable).
- [x] All unit tests (`make test`) are passing.
- [x] Used the same coding conventions as the rest of the project.
- [x] The new code passes pre-commit validations (`pre-commit run --all-files`).
- [x] The PR relates to _only_ one subject with a clear title.
      and description in grammatically correct, complete sentences.

## Approver/Reviewer Checklist

- [ ] Before merge squash related commits.

## Other Information

[Guide to writing a good commit](http://chris.beams.io/posts/git-commit/)

### CODE REVIEW

File: tests/test_context.py
Comment (jfalkenstein) [inline tests/test_context.py:40]: Using the sentinel here doesn't really get you anything. 
```suggestion
        expected = f"{getcwd()}/project_path/to/sceptre"
        assert self.context.project_path.replace(path.sep, "/") == expected
```

**Comment by jfalkenstein** [comment]
I'm triggering integration tests on this now. I'll want to test this out myself before I'm ready to merge it. That won't be until at least this weekend. I'm not sure yet if it makes sense to push this into the coming v4 release yet.

File: tests/test_context.py
Comment (jfalkenstein) [inline tests/test_context.py:24]: Why did you make this change? If this test started failing when you updated the way the path is resolved, then your change cannot work. It will fail in a lot of places where people currently have their CI/CD set to use `--dir project_path/to/sceptre`

Comment (pjsmith404) [reply tests/test_context.py:24]: Sorry, I think this is a misunderstanding on my part of how that test was working.

Maybe it should instead be:
```
expected = f"{getcwd()}/project_path/to/sceptre"
assert self.context.project_path.replace(path.sep, "/") == expected
```

The `--dir` flag does work with the change as it's coercing any given project_path into an absolute path when SceptreContext is initialised

Comment (jfalkenstein) [reply tests/test_context.py:24]: I'll need to do some experimentation on this branch, I think. My own projects use `--dir` quite often.

**Review by jfalkenstein — CHANGES_REQUESTED** [review]
Ok, I just tested this branch out locally with my project that uses the `--dir` cli flag and it works in all three of these situations:

* `sceptre --dir sceptre ...` (path in CWD)
* `sceptre --dir ./sceptre ...` (relative path to CWD)
* `sceptre --dir $PWD/sceptre ...` (absolute path)

So, now that I know it works, I want to see three kinds of tests on this PR then that correspond with each of those three ways to specify the project path. Looking at the code in this PR:
* `test_context_with_path` tests the "path in CWD" approach. I'd like to see that renamed to `test_context_with_path_in_cwd` to make the test case clearer.
* `test_context_with_relative_path` that you have added tests the relative path route.
* I don't currently see a test for handling absolute paths. You should add that.

Once you do that, I'm happy with this PR and I'm ready to merge it!

**Comment by pjsmith404** [comment]
Thanks for the feedback! That should be all covered off now.

**Comment by jfalkenstein** [comment]
Running integration tests on this now...

**Comment by jfalkenstein** [comment]
@zaro0508 can you run integration tests? I'll approve once those pass.

**Review by zaro0508 — APPROVED** [review]

**Pull request merged**
